File: graph/articulation/island.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(int(1e4))

class Land:
    def __init__(self,land_id,coords,always_safe):
        self.land_id = land_id
        self.connection = []
        self.coords = coords
        self.always_safe = always_safe
        self.desc_safe = always_safe #the node has 'always_safe' descendant
        self.safe = always_safe
        self.depth = 0
        self.visit_order = 0
        self.low = 0

    def __repr__(self):
        return f'Land ID: {self.land_id} \n\
connection: {self.connection} \n\
coords: {self.coords} \n\
always_safe: {self.always_safe} \n\
desc_safe: {self.desc_safe} \n\
safe: {self.safe}\n\
depth: {self.depth}\n\
visit_order: {self.visit_order}\n\
low: {self.low}\n'

n,m = map(int,input().split())
grid = [list(input().rstrip()) for _ in range(n)]
grid_visited = [[-1]*m for _ in range(n)]
assign_land_id = 1
lands = [0]

dy = [0,-1,0,1]
dx = [1,0,-1,0]
def assign_land(y,x,always_safe):
    global assign_land_id
    to_visit = [(y,x)]
    coords = []

    while to_visit:
        i,j = to_visit.pop()
        coords.append((i,j))
        grid_visited[i][j] = assign_land_id
        for d in range(4):
            ny = i + dy[d]
            nx = j + dx[d]
            if 0 <= nx < m and 0 <= ny < n and grid_visited[ny][nx] == -1 and grid[ny][nx] == '#':
                to_visit.append((ny,nx))

    lands.append(Land(assign_land_id, coords, always_safe))
    assign_land_id += 1

def find_lands(y,x,always_safe):
    to_visit = [(y,x)]
    found_land = []
    nearby_land = set()
    prev_id = assign_land_id

    while to_visit:
        i,j = to_visit.pop()
        grid_visited[i][j] = 0
        for d in range(4):
            ny = i + dy[d]
            nx = j + dx[d]
            if 0 <= nx < m and 0 <= ny < n:
                if grid_visited[ny][nx] == -1:
                    if grid[ny][nx] == '.':
                        to_visit.append((ny,nx))
                    elif grid[ny][nx] == '#':
                        found_land.append(assign_land_id)
                        assign_land(ny,nx,always_safe)
                if 0 < grid_visited[ny][nx] < prev_id:
                    nearby_land.add(grid_visited[ny][nx])

    for f in found_land:
        for ne in nearby_land:
            lands[f].connection.append(ne)
            lands[ne].connection.append(f)

for i in range(n):
    for j in range(m):
        if grid[i][j] == '.' and grid_visited[i][j] == -1:
            find_lands(i,j,assign_land_id == 1)

time = 1
visit_hist = [0]
def dfs(u,par,depth):
    global time
    lands[u].depth = depth
    lands[u].visit_order = time
    lands[u].low = time
    visit_hist.append(u)
    time += 1

    for v in lands[u].connection:
        if lands[v].visit_order == 0:
            dfs(v,u,depth+1)
            lands[u].low = min(lands[u].low, lands[v].low)
            if lands[v].desc_safe:
                lands[u].desc_safe = True
        elif v != par:
            lands[u].low = min(lands[u].low, lands[v].visit_order)

for i in range(1,len(lands)):
    if lands[i].visit_order == 0:
        dfs(i,-1,1)

for l in lands[1:]:
    if l.desc_safe or l.depth - lands[visit_hist[l.low]].depth >= 2:
        l.safe = True

for l in lands[1:]:
    if l.safe == True:
        for y,x in l.coords:
            grid[y][x] = 'O'
    else:
        for y,x in l.coords:
            grid[y][x] = 'X'

for g in grid:
    print(''.join(g))

File: graph/articulation/island.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <set>
#include <algorithm>

using namespace std;

struct Node {
    int node_id;
    vector<pair<int, int>> coords;
    vector<int> connection;
    bool is_land;
    int depth;
    int visit_order;
    int low;
    bool safe;

    Node(int id, vector<pair<int, int>> coords_, bool is_land_) 
        : node_id(id), coords(coords_), is_land(is_land_), depth(0), visit_order(0), low(0), safe(true) {}
};

int n, m;
vector<vector<char>> grid;
vector<vector<int>> grid_visited;
vector<Node> nodes;
int assign_node_id = 0;

vector<int> dy = {0, -1, 0, 1};
vector<int> dx = {1, 0, -1, 0};
int visit_time = 1;

void assign_node(int y, int x, char icon) {
    stack<pair<int, int>> to_visit;
    vector<pair<int, int>> coords;

    to_visit.push({y, x});

    while (!to_visit.empty()) {
        int i = to_visit.top().first;
        int j = to_visit.top().second;
        to_visit.pop();

        coords.push_back({i, j});
        grid_visited[i][j] = assign_node_id;

        for (int d = 0; d < 4; ++d) {
            int ny = i + dy[d];
            int nx = j + dx[d];
            if (nx >= 0 && nx < m && ny >= 0 && ny < n && grid_visited[ny][nx] == -1 && grid[ny][nx] == icon) {
                to_visit.push({ny, nx});
                grid_visited[ny][nx] = assign_node_id;
            }
        }
    }

    nodes.emplace_back(assign_node_id, coords, icon == '#');
    assign_node_id++;
}

void dfs(int u, int par, int depth) {
    nodes[u].depth = depth;
    nodes[u].visit_order = visit_time;
    nodes[u].low = visit_time;
    visit_time++;

    for (int v : nodes[u].connection) {
        if (nodes[v].visit_order == 0) {
            dfs(v, u, depth + 1);
            nodes[u].low = min(nodes[u].low, nodes[v].low);
        } else if (v != par) {
            nodes[u].low = min(nodes[u].low, nodes[v].visit_order);
        }
    }
}

void mark_unsafe(int start_id) {
    stack<int> to_visit;
    to_visit.push(start_id);

    while (!to_visit.empty()) {
        int p = to_visit.top();
        to_visit.pop();

        nodes[p].safe = false;

        for (int v : nodes[p].connection) {
            if (nodes[v].safe && nodes[v].depth - nodes[p].depth == 1) {
                to_visit.push(v);
            }
        }
    }
}

int main() {
    cin >> n >> m;
    grid.resize(n, vector<char>(m));
    grid_visited.assign(n, vector<int>(m, -1));

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            cin >> grid[i][j];
        }
    }

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            if (grid_visited[i][j] == -1) {
                assign_node(i, j, grid[i][j]);
            }
        }
    }


    for (Node &no : nodes) {
        for (const pair<int,int>&coord : no.coords) {
            int y = coord.first;
            int x = coord.second;
            for (int d = 0; d < 4; ++d) {
                int ny = y + dy[d];
                int nx = x + dx[d];
                if (nx >= 0 && nx < m && ny >= 0 && ny < n && grid_visited[ny][nx] != no.node_id) {
                    no.connection.push_back(grid_visited[ny][nx]);
                }
            }
        }
        set<int> unique_connections(no.connection.begin(), no.connection.end());
        no.connection.assign(unique_connections.begin(), unique_connections.end());
    }

    dfs(0, -1, 1);

    for (Node &no : nodes) {
        if (no.is_land && no.safe == true) {
            for (int near : no.connection) {
                if (nodes[near].depth - no.depth == 1 && nodes[near].low >= no.visit_order && nodes[near].safe == true) {
                    mark_unsafe(near);
                }
            }
        }
    }

    for (Node &no : nodes) {
        if (no.is_land) {
            for (const pair<int,int>&coord : no.coords) {
                int y = coord.first;
                int x = coord.second;
                grid[y][x] = no.safe ? 'O' : 'X';
            }
        }
    }

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            cout << grid[i][j];
        }
        cout << "\n";
    }

    return 0;
}
